pkg/virt-controller/watch/vm.go

```diff
@@ -231,6 +229,10 @@ func (p *authProxy) GetDataSource(namespace, name string) (*cdiv1.DataSource, er
 	return ds, nil
 }
 
+type synchronizer interface {
+	Sync(*virtv1.VirtualMachine, *virtv1.VirtualMachineInstance) (*virtv1.VirtualMachine, error)
+}
+
```

Comment:
Should a synchronize always make a copy, does it returns copy... can you specify this

Reply:
I initially thought about laying out some rules for the usage.
However, I realized this may cause some debate and we need to reach a wider agreement to have such rules. As I see it, this is flexible until we have at least 2 usages.

At its base, the only rules that really matters IMO are these:
- Avoid mutating the input objects (VM and VMI) and return the changes through the returned VM object.
- Avoid changing the `Status` of the returned VM object as it will not be consumed by the main flow.

The enforcement should be by the main flow. Even if the rules are not enforced by the concrete controller, the mainflow will enforce it.

If you still want me to add such rules as formal documentation to the interface, I can. My recommendation is to wait with it to a separate PR which will add another user for the interface.
I'm good either way.

Reply:
I would probably prefer if we don't try to solve this now and we just make an endpoint for the network code for now. The topic is a bit more complex than what is proposed imho. The `synchronizer` first should be called to validate hotplug can be performed, then we would set the condition or continue to call each `synchronizer` based on the validation.

Reply:
I think you are attempting to inticipate how this will look like if and when you will take care of the other hotplug points. As this can go in many directions, I think we better first start with one, then see how it evolves when we add the others one by one.

The advantage to keep this generic is that we encourage other owners of existing codebases and future additions to think in this manner and not add more to the one-big-method.

Reply:
> > I think you are attempting to inticipate how this will look
> 
> I actually argue that the code as of now is trying to forecast how this should look.... 

Yes, it tries to softly forecast how this should look given the existing user of it (the network sync). But this says nothing about the ability to change it per the need of a new user that will come in.

> Therefore I simply suggest to keep this simple and have just one function call instead of the current code block, no interface needed as of now.

The interface is technically needed in order to allow the decoupling between the controller flow and the network implementation.
The idea is to break the dependency, which is the purpose of interfaces.

> > The advantage to keep this generic is that we encourage other owners of existing codebases and future additions to think in this manner and not add more to the one-big-method.
> 
> Simple function has the same effect imho

Functions provide only abstraction layers and do not take care of code architecture concerns like decoupling and focused structured flows.
The sig-network is looking to have this decoupling in place to serve its own code need and encourage others to use the same when introducing new logic.

Do you see now a problem in the current generic suggestion that may break something? Do you think it will make it hard for the owners of this control flow to maintain this part?
I tried to make it as simple as possible so it will not be chalenging to maintain. In fact, it takes out the network logic so it should be easier for this codebase owner to maintain it now.

Reply:
Ok, so would we agree to keep the interface but remove the slice? In its current form it does suggest that it is ok just to add another synchronizer which we already know is not good enough. WDYT?

Reply:
I think it is a reasonable compromise.
I will change it and as a follow up try to add another one in order to return the slice.

Reply:
Done